client/src/components/utils/SASCodeDocument.ts

```diff
@@ -0,0 +1,227 @@
+// Copyright © 2024, SAS Institute Inc., Cary, NC, USA.  All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+interface CodePosition {
+  line: number;
+  character: number;
+}
+
+export interface CodeMetadata {
+  languageId: string;
+  code: string;
+  selectedCode: string;
+  uri?: string;
+  fileName?: string;
+}
+
+export interface CodeSelection {
+  start: CodePosition;
+  end: CodePosition;
+}
+
+export interface SASCodeDocumentOptions {
+  selections?: ReadonlyArray<CodeSelection>;
+  preamble?: string;
+  postamble?: string;
+  outputHtml?: boolean;
+  odsStyle?: string;
+  htmlStyle?: string;
+}
+
+type LineNumber = number;
+type Offset = { lineOffset: number; columnOffset: number };
+
+export class SASCodeDocument {
+  // there may be many selected raw code and they can be not continuous in editor.
+  // this field provides a offset map for selected raw code in wrapped code and in whole raw code.
+  private offsetMap: Map<LineNumber, Offset>;
+
+  constructor(
+    private metadata: CodeMetadata,
+    private options?: SASCodeDocumentOptions,
+  ) {}
+
+  private codeIsEmpty(code: string): boolean {
+    return code.trim() === "";
+  }
+
+  private selectionsAreEmpty(): boolean {
+    return (
+      this.options.selections === undefined ||
+      this.options.selections.length === 0
+    );
+  }
+
+  private generateHtmlStyleOption(): string {
+    const htmlStyle = this.options.htmlStyle ?? "";
+
+    let odsStyle: string;
+    switch (htmlStyle) {
+      case "(auto)":
+        odsStyle = this.options.odsStyle ?? "";
+        break;
+      case "(server default)":
+        odsStyle = "";
+        break;
+      default:
+        odsStyle = htmlStyle;
+        break;
+    }
+
+    return odsStyle ? ` style=${odsStyle}` : "";
+  }
+
+  private wrapCodeWithSASProgramFileName(
+    code: string,
+    fileName: string | undefined,
+  ): string {
+    if (fileName === undefined) {
+      return code;
+    } else {
+      fileName = fileName.replace(/[('")]/g, "%$&");
+      const wrapped =
+        "%let _SASPROGRAMFILE = %nrquote(%nrstr(" + fileName + "));\n" + code;
+      return wrapped;
+    }
+  }
+
+  private wrapCodeWithPreambleAndPostamble(
+    code: string,
+    preamble?: string,
+    postamble?: string,
+  ): string {
+    return (
+      (preamble ? preamble + "\n" : "") +
+      code +
+      (postamble ? "\n" + postamble : "")
+    );
+  }
+
+  private wrapCodeWithOutputHtml(code: string): string {
+    const outputHtml = !!this.options.outputHtml;
+
+    if (outputHtml) {
+      const htmlStyleOption = this.generateHtmlStyleOption();
+      return `title;footnote;ods _all_ close;
+ods graphics on;
+ods html5${htmlStyleOption} options(bitmap_mode='inline' svg_mode='inline');
```

Comment:
This will need changes from main, here - https://github.com/sassoftware/vscode-sas-extension/blob/1c38235d2de440744669ab081d8d639be610c481/client/src/components/utils/sasCode.ts#L52

Reply:
Done